### PR TITLE
Pushing the build image to a Docker image repository

### PR DESCRIPTION
**Pull request by arschles**
By pushing the build image to a repo and `docker pull`-ing it, developers no longer need to build it locally. This patch also modifies the Makefile to push and use the new build image. 

See https://quay.io/repository/kubernetes-service-catalog/build-image for the build image's location.

Fixes https://github.com/kubernetes-incubator/service-catalog/issues/778

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,5 @@ contrib/build/*/tmp/*
 .pkg
 .kube
 .var
+build-image
+push-build-image
```
Comment (duglin) [inline .gitignore]: can we put a dot in front of these to avoid cluttering our dir with empty files?
We can still have nice build targets if we want/need:
```
.PHONY: build-image

build-image: .build-image
    ... 
```

Comment (arschles) [inline .gitignore]: sure

**Comment by duglin** [comment]
LGTM if CI is happy

**Comment by arschles** [comment]
Failing [in Jenkins](https://service-catalog-jenkins.appspot.com/job/service-catalog-PR-testing2/505/console) because of this line: 

```console
[Build & Unit/Integration Tests] docker pull gcr.io/service-catalog-jenkins-jobs/catalog/build-image:canary
```

File: contrib/hack/start-server.sh [hunk 2]
```diff
@@ -40,7 +40,7 @@ docker run -d --name apiserver \
 	-e KUBERNETES_SERVICE_PORT=6443 \
 	--privileged \
 	--net container:etcd-svc-cat \
-	scbuildimage \
+	quay.io/kubernetes-service-catalog/build-image:canary \
```
Comment (duglin) [inline contrib/hack/start-server.sh]: since we allow people to change the BUILD_IMAGE_REGISTRY value during build time, for consistency, should we also pick up that value here?  But default to quay?

File: Makefile [hunk 3]
```diff
@@ -292,7 +305,7 @@ controller-manager-image: build/controller-manager/Dockerfile $(BINDIR)/controll
 
 # Push our Docker Images to a registry
 ######################################
-push: user-broker-push controller-manager-push apiserver-push
+push: user-broker-push controller-manager-push apiserver-push push-build-image
```
Comment (duglin) [inline Makefile]: add dot before push-build-image to fix CI error.

File: contrib/hack/start-server.sh (hunk 2 above)
Comment (arschles) [inline contrib/hack/start-server.sh]: we could, but I'd like to leave that to a different PR

**Comment by krancour** [comment]
I see a potential problem here...

If the build image is mutable, then we can really quite easily find ourselves in a scenario where we're trying to compile an old version of the code or a specific branch of this repo and the build image isn't suitable for doing so.

This is at least somewhat related to my [previous question](https://github.com/kubernetes-incubator/service-catalog/issues/778#issuecomment-298362543) about this.

**Comment by arschles** [comment]
@krancour I'm not sure I understand the problem here. the `.init` target can pull down a new image if necessary, and developers can change the build image they intend to use. We can `.init` to also build the image if that addresses the issue?

**Comment by MHBauer** [comment]
Is there an escape hatch if I want to build the image myself, or not pull the upstream image?

**Comment by MHBauer** [comment]
I'm not opposed, I just don't want to pull this over the network.
```
scbuildimage                                             latest                        5c4a06fb5527        2 weeks ago          731.4 MB
```

**Comment by arschles** [comment]
@MHBauer if you put `BUILD_IMAGE_MUTABLE=scbuildimage` in your env, then run `make build-image`, you can execute any other `make` command without pulling from the network

**Comment by krancour** [comment]
@arschles 

> the .init target can pull down a new image if necessary

By default, it will always pull down `quay.io/kubernetes-service-catalog/build-image:canary`, which is built from the head of master.

Suppose I am trying to build "master - n"? The build image from the head of master may not be suitable for that.

I recognize that I could elect to build the build image myself. But I might not always know that there's a need to do so. _Before_ this PR, even if it was a bit slow, I knew I could _reliably_ `make clean` and start fresh with confidence that the correct build image would be built for me as a pre-requisite for everything else. Now, unless I am missing something, I would have to _know_ I needed a custom build of the build image and I'd have to manually execute some steps to accommodate that.

If I'm reading this all wrong, then perhaps this PR can benefit from some docs that explain the workflow for the average case _and_ for the cases where you're building something other than the head of master.

**Comment by krancour** [comment]
To be clear, I simply want to avoid:

1. Having to think about this.
2. Creating any confusion for new contributors.

File: Makefile [hunk 4]
```diff
@@ -291,8 +304,10 @@ controller-manager-image: build/controller-manager/Dockerfile $(BINDIR)/controll
 	rm -rf build/controller-manager/tmp
 
 # Push our Docker Images to a registry
+# note: .push-build-image should be the first dependency, because all others rely on the build
+# image
 ######################################
-push: user-broker-push controller-manager-push apiserver-push
+push: .push-build-image user-broker-push controller-manager-push apiserver-push
```
Comment (kibbles-n-bytes) [inline Makefile]: I don't think pushing the build image should be lumped in with the rest of the pushes. They're two totally different actions.

`user-broker-push`, `controller-manager-push`, `apiserver-push`: these are pushing images of freshly compiled catalog binaries to your personal registry. `push-build-image`, on the other hand, is updating a shared image on a shared registry that is used during the build process. Adding `push-build-image` to this action makes it so you have to specifically name the user-broker, controller-manager, and apiserver actions if you just want to do your personal registry pushes.

If someone wants to push the build image, they can just explicitly call that command.

**Comment by MHBauer** [comment]
As long as it exists, I'll probably end up leaving that var set.

What problem are we trying to avoid with this?

**Comment by arschles** [comment]
@krancour not sure what exactly the average case is, so I'll assume it's building from the HEAD of master. In that case, nobody has to think about this and there won't be any confusion. In other cases like building from `master - n`, we suffer from not versioning the build image.

If we did that, then we could default the build image name to `quay.io/kubernetes-service-catalog/build-image:${VERSION}` and the builds would "just work" from this point onward. 

However, builds from points before this version would require the developer to manually specify the build image as `canary` because versioned images for those SHAs wouldn't exist.

Thoughts?

**Comment by arschles** [comment]
Marking this as in progress, because I need to think through some longer term implications of this change

**Comment by MHBauer** [comment]
documenting the escape hatch usage of BUILD_IMAGE_MUTABLE= needs to be done.

**Comment by duglin** [comment]
if it possible to make a master build of travis (or jekins) update the official quay build image if things are successful?  Then we wouldn't need to worry about what @krancour said here: https://github.com/kubernetes-incubator/service-catalog/issues/778#issuecomment-298362543

**Comment by arschles** [comment]
There are still problems with this implementation and open questions on the issue itself, so I am closing this PR w/o merge. Please see https://github.com/kubernetes-incubator/service-catalog/issues/778#issuecomment-299333218 for more detail
